### PR TITLE
refactor(storage): read object to use rest library

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197)

### CODE REVIEW

File: google/cloud/internal/curl_impl.cc [hunk 1]
```diff
@@ -625,6 +630,7 @@ StatusOr<std::size_t> CurlImpl::ReadImpl(absl::Span<char> output) {
     return bytes_read;
   }
   TRACE_STATE() << ", http code=" << http_code_ << "\n";
+  received_headers_.emplace(":curl-peer", handle_.GetPeer());
```
Comment (coryan) [inline google/cloud/internal/curl_impl.cc:633]: I recently removed this as part of #9162, please do not introduce it again.

File: google/cloud/storage/internal/object_requests.cc [hunk 2]
```diff
@@ -194,32 +194,38 @@ bool ReadObjectRangeRequest::RequiresRangeHeader() const {
   return RequiresNoCache();
 }
 
-std::string ReadObjectRangeRequest::RangeHeader() const {
+std::string ReadObjectRangeRequest::RangeHeaderValue() const {
```
Comment (coryan) [inline google/cloud/storage/internal/object_requests.cc:197]: nit: this could have been a separate PR.

File: google/cloud/storage/client.cc [hunk 3]
```diff
@@ -55,9 +56,47 @@ std::shared_ptr<internal::RawClient> Client::CreateDefaultInternalClient(
...
+    if (opts.has<EnableCurlSslLockingOption>()) {
+      rest_opts.set<rest::EnableCurlSslLockingOption>(
+          opts.get<EnableCurlSslLockingOption>());
+    }
+    if (opts.has<EnableCurlSigpipeHandlerOption>()) {
+      rest_opts.set<rest::EnableCurlSigpipeHandlerOption>(
+          opts.get<EnableCurlSigpipeHandlerOption>());
+    }
+
     return CreateDefaultInternalClient(opts,
```
Comment (coryan) [inline google/cloud/storage/client.cc:100]: Did you intend to use `rest_opts` somewhere?

File: google/cloud/storage/tests/object_media_integration_test.cc [hunk 4]
```diff
@@ -609,6 +609,13 @@ TEST_F(ObjectMediaIntegrationTest, ConnectionFailureUploadFile) {
 
 TEST_F(ObjectMediaIntegrationTest, StreamingReadTimeout) {
   if (!UsingEmulator()) GTEST_SKIP();
+  // TODO(9201): RestClient does not encounter the CURL error [28] Timeout
```
Comment (coryan) [inline google/cloud/storage/tests/object_media_integration_test.cc:612]: I think my previous comment about the unused `rest_opts` explains this.

File: google/cloud/storage/internal/rest_client.cc [hunk 5]
```diff
@@ -574,9 +575,77 @@ StatusOr<ObjectMetadata> RestClient::GetObjectMetadata(
...
+    builder.AddHeader("Range", request.RangeHeaderValue());
+  }
+  if (request.RequiresNoCache()) {
+    builder.AddHeader("Cache-Control", "no-transform");
+  }
+
+  auto response = storage_rest_client_->Get(std::move(builder).BuildRequest());
+  if (!response.ok()) return response.status();
+
+  return std::unique_ptr<ObjectReadSource>(
```
Comment (coryan) [inline google/cloud/storage/internal/rest_client.cc:647]: Why no `absl::make_unique<>` here?

**Review by coryan — CHANGES_REQUESTED** [review]
A few small nits, and I think you can remove a TODO. Looking good overall.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197)* status: 0 of 7 files reviewed, 5 unresolved discussions (waiting on @coryan and @scotthart)

___
*[`google/cloud/internal/curl_impl.cc` line 633 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fzdv-r1-633:-N4-XmzE6NTwhNqry0_9:b-keeub7) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/internal/curl_impl.cc#L633)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I recently removed this as part of #9162, please do not introduce it again.
</blockquote></details>

So I initially rebased without adding the curl-peer header, but that caused a failure in ObjectReadHeadersIntegrationTest_NoDuplicatePeers, so I added it back in, which results in the test passing.
___
*[`google/cloud/storage/client.cc` line 100 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fzdw-r1-100:-N4-XhtU61sQ_7n7fbpj:b-834n6i) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/storage/client.cc#L100)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Did you intend to use `rest_opts` somewhere?
</blockquote></details>

Oops.
___
*[`google/cloud/storage/internal/rest_client.cc` line 647 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fzdz-r1-647:-N4-XPjl2rVONiiYkwVZ:b-q26htx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/storage/internal/rest_client.cc#L647)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Why no `absl::make_unique<>` here?
</blockquote></details>

Because the return type is `StatusOr<std::unique_ptr<ObjectReadSource>>` and `absl::make_unique<RestObjectReadSource>()` covariant issues.
___
*[`google/cloud/storage/tests/object_media_integration_test.cc` line 612 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fze0-r1-612:-N4-Y8HH6rLk_zKhQTUH:bdafnxm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/storage/tests/object_media_integration_test.cc#L612)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think my previous comment about the unused `rest_opts` explains this.
</blockquote></details>

Indeed it does, nice catch.

**Review by coryan — APPROVED** [review]
> Reviewed 5 of 7 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @scotthart)

___
*[`google/cloud/internal/curl_impl.cc` line 633 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fzdv-r1-633:-N4-amQR9kqVVIrBTw6y:b-koshnl) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/internal/curl_impl.cc#L633)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

So I initially rebased without adding the curl-peer header, but that caused a failure in ObjectReadHeadersIntegrationTest_NoDuplicatePeers, so I added it back in, which results in the test passing.
</blockquote></details>

Ack.  Well if the test passes we are Okay.  Sorry I introduced that bug.
___
*[`google/cloud/storage/internal/rest_client.cc` line 647 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N4-8rlWpMqzOCo0Fzdz-r1-647:-N4-bDcD4GJGw3vJ1w86:b8v2pct) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/273600d14bd948e4b94d77cc63fd7371aae1dad6/google/cloud/storage/internal/rest_client.cc#L647)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Because the return type is `StatusOr<std::unique_ptr<ObjectReadSource>>` and `absl::make_unique<RestObjectReadSource>()` covariant issues.
</blockquote></details>

Ack, thanks.

File: google/cloud/internal/curl_options.h [hunk 6]
```diff
@@ -62,6 +62,8 @@ struct CAPathOption {
  * value may result in very few connections if your application does not need
  * them.
  *
+ * @note Setting this value to 0 results disables connection pooling.
```
Comment (devbww) [inline google/cloud/internal/curl_options.h:65]: s/results //

File: google/cloud/storage/options.h [hunk 7]
```diff
@@ -102,6 +102,8 @@ struct ProjectIdOption {
  * value may result in very few connections if your application does not need
  * them.
  *
+ * @note Setting this value to 0 results disables connection pooling.
```
Comment (devbww) [inline google/cloud/storage/options.h:105]: Ditto.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197)* status: 7 of 10 files reviewed, 2 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/internal/curl_options.h` line 65 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N43LtZlBuDn4oCGs1Pb-r4-65:-N43RADs5CgdRL7cA15B:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/24724b93058a5af43555b01b018de1ff6ee616a2/google/cloud/internal/curl_options.h#L65)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

s/results //
</blockquote></details>

Done.
___
*[`google/cloud/storage/options.h` line 105 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197#-N43LtZlBuDn4oCGs1Pc-r4-105:-N43RAcXBood5g6OVzi8:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/24724b93058a5af43555b01b018de1ff6ee616a2/google/cloud/storage/options.h#L105)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 1 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9197)* status: all files reviewed, 2 unresolved discussions (waiting on @devbww)

**Pull request merged**